File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/data/database/LockRealmDataStore.java
package com.lattis.ellipse.data.database;

import com.lattis.ellipse.data.database.base.RealmObservable;
import com.lattis.ellipse.data.database.mapper.RealmLockMapper;
import com.lattis.ellipse.data.database.model.RealmLock;
import com.lattis.ellipse.domain.model.Lock;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.functions.Function;
import io.realm.Realm;
import io.realm.RealmConfiguration;

public class LockRealmDataStore {

    private RealmLockMapper lockMapper;
    private RealmConfiguration realmConfiguration;

    @Inject
    LockRealmDataStore(RealmConfiguration realmConfiguration, RealmLockMapper lockMapper) {
        this.lockMapper = lockMapper;
        this.realmConfiguration = realmConfiguration;
    }


    public Observable<Lock> createOrUpdateLock(Lock lock) {
        return RealmObservable.object(
                new Function<Realm, RealmLock>() {
                    @Override
                    public RealmLock apply(Realm realm) {
                        return realm.copyToRealmOrUpdate(lockMapper.mapIn(lock));
                    }
                })
                .map(new Function<RealmLock, Lock>() {
                    @Override
                    public Lock apply(RealmLock realmlock) {
                        return lockMapper.mapOut(realmlock);
                    }
                });
    }

    public Observable<Lock> getLock() {
        return RealmObservable.object(
                realm -> realm.where(RealmLock.class)
                        .findFirst())
                .map(realmLock -> lockMapper.mapOut(realmLock));
    }


    public Observable<Boolean> deleteLock() {
        return RealmObservable.deleteObject(
                realm -> {
                    RealmLock realmLock = realm.where(RealmLock.class)
                            .findFirst();
                    if(realmLock != null){
                        realmLock.deleteFromRealm();
                        return true;
                    }
                    return false;
                });
    }



}

File: velolabs_repos/lattis_android_v2/data/src/main/java/com/lattis/data/database/store/LockRealmDataStore.kt
package com.lattis.data.database.store

import com.lattis.data.database.base.RealmObservable
import com.lattis.data.database.base.RealmObservable.`object`
import com.lattis.data.database.base.RealmObservable.deleteObject
import com.lattis.data.database.mapper.RealmLockMapper
import com.lattis.data.database.model.RealmLock
import com.lattis.domain.models.Lock
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.functions.Function
import io.realm.Realm
import io.realm.RealmConfiguration
import io.realm.RealmObject
import javax.inject.Inject

class LockRealmDataStore @Inject internal constructor(
    private val realmConfiguration: RealmConfiguration,
    private val lockMapper: RealmLockMapper
) {
    fun createOrUpdateLock(lock: Lock): Observable<Lock> {
        return RealmObservable. `object`(
            Function<Realm, RealmLock> { realm ->
                realm.copyToRealmOrUpdate(
                    lockMapper.mapIn(
                        lock!!
                    )
                )
            })
            .map( {realmlock->
                     lockMapper.mapOut(realmlock!!)
                }
            )
    }

    fun getLock(): Observable<Lock>
    {
        return RealmObservable.`object`<RealmLock>(
            Function { realm->
                val realmlock = realm.where(RealmLock::class.java).findFirst()
                if (realmlock != null) realmlock else RealmLock()
            }
        ).map { realmLock -> lockMapper.mapOut(realmLock) }
    }

    fun deleteLock(): Observable<Boolean> {
        return deleteObject(
            Function { realm: Realm ->
                val realmLock = realm.where(RealmLock::class.java)
                    .findFirst()
                if (realmLock != null) {
                    realmLock.deleteFromRealm()
                    true
                }
                false
            }
        )
    }

}